### ISSUE TITLE
Multiple responses per request

### ISSUE THREAD

**Issue by Ullaakut** [nanomsg/mangos#228]
Hi,

It's not clear to me whether Mangos supports sending multiple messages over one socket after receiving one single request. I was under the impression that using a `OpenContext` on a socket that uses a protocol that supports it, it would be possible to achieve something like:

```go
ctx, _ := sock.OpenContext()

req, _ = ctx.Recv()

// imaginary function that returns a response in multiple chunks from a request
resp := getResponseChunks(req)

for _, chunk := range resp {
	ctx.Send(chunk)
}
```

But when doing so using a `rep` socket over `tcp`, the second call to `ctx.Send` returns an `incorrect protocol state` error, so I imagine that it's not possible.

I haven't found detailed documentation on which protocol supports which operations, using contexts, so I might be doing something wrong. My use case is that I need to send a large amount of data (potentially multiple gigabytes) over the socket, so I want to split it into multiple responses. Maybe there's another, better way to achieve that using mangos?

**Comment by gdamore** [comment]
Each context has its own state machine.  So the way to do this is to open a separate context for each message you want to send in parallel. Contexts are fairly cheap and as long as you don't create too many of them they shouldn't have any noticeable impact.  (Too many is probably in the thousands.)

**Comment by Ullaakut** [comment]
@gdamore Thanks for the quick answer. So how would one go about creating multiple contexts from one open connection to a client?

Something as simple as this, to use my previous example?

```go

req, _ = reqSock.Recv()

// imaginary function that returns a response in multiple chunks from a request
resp := getResponseChunks(req)

for _, chunk := range resp {
	ctx, _ := sock.OpenContext()
	ctx.Send(chunk)
}
```

**Comment by gdamore** [comment]
yes, that's pretty much it.

**Comment by gdamore** [comment]
Actually, that's not right.

You should open the context *before* doing receive.

You want to start a bunch of them in parallel.  That determines the amount of parallelization you can deal with.

**Comment by gdamore** [comment]
```
for i := 0; i < nparallel; i++ {
    ctx, _ := sock.OpenContext()
    go func() {
    req, _ := reqSock.Recv()
    // do work
    ctx.Send(reply)
    }()
}
```

You could even do the above with an inner loop:

```
for i := 0; i < nparallel; i++ {
    ctx, _ := sock.OpenContext()
    go func() { 
        for {
            req, _ := reqSock.Recv()
            // do work
           ctx.Send(reply)
    }()
}
```

**Comment by gdamore** [comment]
Note that the REQ/REP protocol requires exactly one reply per request.  You cannot send multiple replies to a request.

**Comment by Ullaakut** [comment]
@gdamore Aha, that was not 100% clear to me. Thanks for your explanations :)

Is there a supported protocol in Mangos that would allow multiple responses for one specific request? If not, I could always manage to do it with a websocket but it would be nice to have something that natively supports it rather than to use a pub/sub system to replicate it.

**Comment by gdamore** [comment]
Unfortunately, no.  If you had multiple contexts open, you could use surveyor, but this would be very awkward and you'd have no control over the order of handling & delivery.  The need for such a pattern has never arisen before this.

**Comment by Ullaakut** [comment]
Mh that is a shame but I can't say I'm surprised. If you're curious, my use case is the following:

I am building a system that follows the live state of another system and exposes its history as an API. In order for this to work with a running network, the following is needed:

* When connecting to the live system, my nodes need to ask the live system for its history from the start to its current state. This is done via a req/rep socket to query chunks of updates in ascending order.
* Once this bootstrapping is done and the followers are at the same state as the live nodes, they apply updates that are published by the live nodes on a pub socket. Those updates are cached until the bootstrapping process is finished.

Currently however, this bootstrapping process is very slow, because for each chunk, the followers need to ask for a certain range of updates (for example, "I want updates 1520 to 1550"), then get a response, ask for another one, etc. I thought it would be more efficient to somehow just ask for all updates from 0 to 3 000 000 (which can represent multiple gigabytes of data in my case) and receive them chunk by chunk without having do make any further request.

**Comment by gdamore** [comment]
This use case makes sense.  You can actually bulk all the updates together in a single reply -- which is what I would do.  mangos can use large messages.  If you are uncomfortable with a multigigabyte message, you could also just send in batches -- and have the request request the next batch -- this is sort of how pagination APIs typically work.

I think just widening the window of your requests (ask for 1000 at a time or 10000 at a time, or a size limit based on e.g. 10 MB max or something) will go fast, and be easily supportable even with just single reply per request.  The trick is to make sure that you send enough updates at once (in one message) so that the round-trip overhead is a minimum percentage.

You can calculate this in size btw... For example if your round trip time is 10 msec, then to minimize the round trip overhead you'd want to send enough data to fill the pipe for 100ms or more.  If you have a Gigabit link, it should take around a second to send 128MB... which means if you send 128MB in a single request your overhead only accounts for 1%.  (You can use smaller messages, with larger overhead percentage -- 12MB means 10% round trip overhead assuming the same link speed and round trip time.)

**Comment by gdamore** [comment]
Btw, to really get the most performance you might want to have multiple parallel streams.  Open up multiple dialers, and break the request up in to batches by dialer.  This is progressively more important the longer the round trip (TCP) time, because TCP single stream performance is pretty crummy especially over long haul or lossy networks.   Experimentally with 10GB links over a cross country network we found we needed like 8 parallel streams to get good performance.

**Comment by ioannist** [comment]
how come there is no context used in the example file? https://github.com/nanomsg/mangos/blob/master/examples/reqrep/reqrep.go

**Comment by gdamore** [comment]
The example was created before the context API was created.  You don't have to use contexts if you are only going to have simple synchronous uses.
